libp2p-comit/src/protocol.rs

```diff
@@ -48,3 +50,173 @@ impl OutboundUpgrade<NegotiatedSubstream> for Config {
         future::ok(framed)
     }
 }
+
+/// Represents a prototype for an upgrade to handle the secret hash sharing
+/// protocol.
+///
+/// The protocol works in the following way:
+///
+/// - Dialer (Bob) reads the `SecretHashInfo` from the substream.
+/// - Listener (Alice) writes the `SecretHashInfo` to the substream.
+#[derive(Debug, Clone)]
+pub struct SecretHashProtocolConfig;
+
+/// Data sent to peer in secret hash protocol.
+#[derive(Deserialize, Debug, Serialize)]
+pub struct SecretHashInfo {
+    swap_id: String,     // FIXME: Use SwapId type
+    secret_hash: String, // FIXME: Use secret type from cnd?
+}
+
+/// The substream on which a reply is expected to be sent.
+pub struct ReplySubstream<T> {
+    inner: T,
+}
+
+impl<T> fmt::Debug for ReplySubstream<T> {
+    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
+        f.debug_tuple("ReplySubstream").finish()
+    }
+}
+
+impl<T> ReplySubstream<T>
+where
+    T: AsyncWrite + Unpin,
+{
+    /// Sends back the requested information on the substream.
+    ///
+    /// Consumes the substream, returning a `ReplyFuture` that resolves
+    /// when the reply has been sent on the underlying connection.
+    pub fn send(mut self, info: SecretHashInfo) -> impl Future<Output = Result<(), io::Error>> {
+        debug!("Sending secret hash info to peer");
+
+        let message = serde_json::to_vec(&info).unwrap();
+
+        async move { upgrade::write_one(&mut self.inner, &message).await }
+    }
+}
+
+impl UpgradeInfo for SecretHashProtocolConfig {
+    type Info = &'static [u8];
+    type InfoIter = iter::Once<Self::Info>;
+
+    fn protocol_info(&self) -> Self::InfoIter {
+        iter::once(b"/comit/swap/secret_hash/1.0.0") // FIXME: add a helper
+                                                     // function to generate
+                                                     // this.
+    }
+}
+
+impl<C> InboundUpgrade<C> for SecretHashProtocolConfig
+where
+    C: AsyncRead + AsyncWrite + Unpin,
+{
+    type Output = ReplySubstream<C>;
+    type Error = io::Error;
+    type Future = future::Ready<Result<Self::Output, io::Error>>;
+
+    fn upgrade_inbound(self, socket: C, _: Self::Info) -> Self::Future {
+        trace!("Upgrading inbound connection");
+        future::ok(ReplySubstream { inner: socket })
+    }
+}
+
+impl<C> OutboundUpgrade<C> for SecretHashProtocolConfig
+where
+    C: AsyncRead + AsyncWrite + Unpin + Send + 'static,
+{
+    type Output = SecretHashInfo;
+    type Error = upgrade::ReadOneError;
+    type Future = Pin<Box<dyn Future<Output = Result<Self::Output, Self::Error>> + Send>>;
+
+    fn upgrade_outbound(self, mut socket: C, _: Self::Info) -> Self::Future {
+        trace!("Upgrading outbound connection");
+        Box::pin(async move {
+            let msg = upgrade::read_one(&mut socket, 4096).await?; // FIXME: how much data to read?
```

Comment:
Given we know the size of the message (UUID for swapid and secret hash are always the same size, we can calculate this upfront. What is to be investigated is whether this is a buffer or actually a fixed amount. We cannot read more than the sender is trying to send.

Reply:
Note: @thomaseizinger discovered that the size is just a maximum to prevent DOS attacks against a node.  Using 1024.